mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ExecutorStateManagerImpl.java

```diff
@@ -318,29 +332,71 @@ private Optional<Map<TaskExecutorID, TaskExecutorState>> findBestFitFor(TaskExec
             return Optional.empty();
         }
 
-        return Optional.of(
-            this.executorByCores.get(targetCoreCount)
-                .descendingSet()
-                .stream()
-                .filter(teHolder -> {
-                    if (!this.taskExecutorStateMap.containsKey(teHolder.getId())) {
-                        return false;
-                    }
+        Map<TaskExecutorID, TaskExecutorState> availableTEs = findAvailableBestFitTaskExecutors(schedulingConstraints, targetCoreCount, currentBestFit, numWorkers);
+        return availableTEs.size() < numWorkers ? Optional.empty() : Optional.of(availableTEs);
+    }
 
-                    if (currentBestFit.contains(teHolder.getId())) {
-                        return false;
+    /**
+     * Identifies 'numWorkers' Task Executors that are available, not currently used, and comply with the given scheduling constraints.
+     * From these, the method selects the ones with the highest fitness score currently based on cpu and memory.
+     *
+     * @param schedulingConstraints The requested scheduling constraints to evaluate each TaskExecutor.
+     * @param targetCoreCount The target core count to select TaskExecutors from.
+     * @param currentBestFit Previously selected best fit TaskExecutors to exclude from consideration.
+     * @param numWorkers The desired number of workers in the final selection.
+     *
+     * @return A map of available TaskExecutorIDs to their TaskExecutorState that best fit the requested scheduling constraints.
+     * The number of entries is limited to `numWorkers`. If no TaskExecutor satisfies the scheduling constraints, this would be an empty map.
+     */
+    private Map<TaskExecutorID, TaskExecutorState> findAvailableBestFitTaskExecutors(SchedulingConstraints schedulingConstraints, Double targetCoreCount, BestFit currentBestFit, Integer numWorkers) {
+        double bestFitnessScore = 0;
+        List<TaskExecutorHolder> bestFittingTEs = new ArrayList<>();
+
+        for (TaskExecutorHolder teHolder: this.executorByCores.get(targetCoreCount).descendingSet()) {
```

Comment:
it seems rather in-efficient to track all TEs under same core count then need to re-categorize based on attributes. I feel like the previous approach of tracking via a SKU variant ID is more efficient/safer.

Reply:
fair point. I've updated the logic to index by a group key so we shouldn't have this problem.